crates/subspace-farmer/src/piece_cache.rs

```diff
@@ -0,0 +1,666 @@
+use crate::single_disk_plot::piece_cache::{DiskPieceCache, Offset};
+use crate::utils::AsyncJoinOnDrop;
+use crate::NodeClient;
+use futures::lock::Mutex;
+use futures::{select, FutureExt, StreamExt};
+use parking_lot::RwLock;
+use rayon::prelude::*;
+use std::collections::HashMap;
+use std::mem;
+use std::sync::Arc;
+use subspace_core_primitives::{Piece, PieceIndex, SegmentIndex};
+use subspace_farmer_components::plotting::{PieceGetter, PieceGetterRetryPolicy};
+use subspace_networking::libp2p::kad::RecordKey;
+use subspace_networking::libp2p::PeerId;
+use subspace_networking::utils::multihash::ToMultihash;
+use subspace_networking::{KeyWrapper, UniqueRecordBinaryHeap};
+use tokio::sync::mpsc;
+use tracing::{debug, error, info, trace, warn};
+
+const WORKER_CHANNEL_CAPACITY: usize = 100;
+/// Make caches available as they are building without waiting for the initialization to finish,
+/// this number defines an interval in pieces after which cache is updated
+const INTERMEDIATE_CACHE_UPDATE_INTERVAL: usize = 100;
+
+#[derive(Debug, Clone)]
+struct DiskPieceCacheState {
+    stored_pieces: HashMap<RecordKey, Offset>,
+    free_offsets: Vec<Offset>,
+    backend: DiskPieceCache,
+}
+
+#[derive(Debug)]
+enum WorkerCommand {
+    ReplaceBackingCaches { new_caches: Vec<DiskPieceCache> },
+    ForgetKey { key: RecordKey },
+}
+
+#[derive(Debug)]
+struct CacheWorkerState {
+    heap: UniqueRecordBinaryHeap<KeyWrapper<PieceIndex>>,
+    last_segment_index: SegmentIndex,
+}
+
+/// Cache worker used to drive the cache
+#[must_use = "Cache will not work unless its worker is running"]
+pub struct CacheWorker<NC> {
+    peer_id: PeerId,
+    node_client: NC,
+    /// It is important to always lock caches AFTER worker state in order to avoid deadlock!
+    caches: Arc<RwLock<Vec<DiskPieceCacheState>>>,
+    worker_receiver: Option<mpsc::Receiver<WorkerCommand>>,
+}
+
+impl<NC> CacheWorker<NC>
+where
+    NC: NodeClient,
+{
+    /// Run the cache worker with provided piece getter
+    pub async fn run<PG>(mut self, piece_getter: PG)
+    where
+        PG: PieceGetter,
+    {
+        // Limit is dynamically set later
+        // It is important to always lock worker state BEFORE caches in order to avoid deadlock!
+        let worker_state = Mutex::new(CacheWorkerState {
+            heap: UniqueRecordBinaryHeap::new(self.peer_id, 0),
+            last_segment_index: SegmentIndex::ZERO,
+        });
+
+        let Some(mut worker_receiver) = self.worker_receiver.take() else {
+            error!("Worker receiver was not set, this is an implementation bug");
+            return;
```

Comment:
Why `return` and not panic or `return Err`? It seems that it goes to `run_future_in_dedicated_thread` which doesn't provide fatal error handling. Did I miss it?

Reply:
This should never actually happen, so yes, it could be a panic. Just didn't feel like panicking at the time for some reason, will change.